cpp/include/cugraph/sampling_functions.hpp

```diff
@@ -240,6 +244,116 @@ biased_neighbor_sample(
+ * the label output has values.  This will also result in the output being sorted by vertex label.
+ * The offsets array in the return will be a CSR-style offsets array to identify the beginning of
+ * each label range in the data.  `labels.size() == (offsets.size() - 1)`.
+ *
+ * If @p starting_vertex_labels is specified and @p label_to_output_comm_rank is specified then the
+ * label output has values.  This will also result in the output being sorted by vertex label.  The
+ * offsets array in the return will be a CSR-style offsets array to identify the beginning of each
+ * label range in the data.  `labels.size() == (offsets.size() - 1)`.  Additionally, the data will
+ * be shuffled so that all data with a particular label will be on the specified rank.
+ *
+ * @tparam vertex_t Type of vertex identifiers. Needs to be an integral type.
+ * @tparam edge_t Type of edge identifiers. Needs to be an integral type.
+ * @tparam weight_t Type of edge weights. Needs to be a floating point type.
+ * @tparam edge_type_t Type of edge type. Needs to be an integral type.
+ * @tparam label_t Type of label. Needs to be an integral type.
+ * @tparam store_transposed Flag indicating whether sources (if false) or destinations (if
+ * true) are major indices
+ * @tparam multi_gpu Flag indicating whether template instantiation should target single-GPU (false)
+ * @param handle RAFT handle object to encapsulate resources (e.g. CUDA stream, communicator, and
+ * handles to various CUDA libraries) to run graph algorithms.
+ * * @param rng_state A pre-initialized raft::RngState object for generating random numbers
+ * @param graph_view Graph View object to generate NBR Sampling on.
+ * @param edge_weight_view Optional view object holding edge weights for @p graph_view.
+ * @param edge_id_view Optional view object holding edge ids for @p graph_view.
+ * @param edge_type_view Optional view object holding edge types for @p graph_view.
+ * @param edge_bias_view Optional view object holding edge biases (to be used in biased sampling) for @p
+ * graph_view. Bias values should be non-negative and the sum of edge bias values from any vertex
+ * should not exceed std::numeric_limits<bias_t>::max(). 0 bias value indicates that the
+ * corresponding edge can never be selected.
+ * @param starting_vertices Device span of starting vertex IDs for the sampling.
+ * In a multi-gpu context the starting vertices should be local to this GPU.
+ * @param starting_vertex_labels Optional device span of labels associted with each starting vertex
+ * for the sampling.
+ * @param label_to_output_comm_rank Optional tuple of device spans mapping label to a particular
+ * output rank.  Element 0 of the tuple identifes the label, Element 1 of the tuple identifies the
+ * output rank.  The label span must be sorted in ascending order.
+ * @param fan_out Host span defining branching out (fan-out) degree per source vertex for each
```

Comment:
We need something in the documentation to indicate that either `fan_out` or `heterogeneous_fan_out` need to be specified... but not both.  Perhaps something that mentions that `fan_out` uses the same fanout value for each type and that `heterogeneous_fan_out` uses a different fanout value for each type.

Probably need to discuss what happens if you don't specify all types.  That is, if there are 5 unique types in the graph and you only specify fanout values for 3 of them... what happens.

Reply:
> Probably need to discuss what happens if you don't specify all types. That is, if there are 5 unique types in the graph and you only specify fanout values for 3 of them... what happens.

I think in this case we have an error right? we assume internally the edge types' range to be [0, # edge types). Hence if a user doesn't provide the corresponding set of fan_out values, there is an issue. Essentially the `fan_out_offsets` must to be `n + 1` where n is the # edge types.

Reply:
Below is the description I have so far

```
 * @brief Neighborhood Sampling.
 *
 * This function traverses from a set of starting vertices, traversing outgoing edges and
 * randomly selects (with edge biases or not) from these outgoing neighbors to extract a subgraph.
 * When branching out to select outgoing neighbors, either fan_out or heterogeneous_fan_out must
 * be provided but not both.

 * @param fan_out Host span defining branching out (fan-out) degree per source vertex for each
 * level. When fan_out is provided, the sampling method uses the same fanout value for each type.
 * @param heterogeneous_fan_out Tuple of host spans defining branching out (fan-out) degree per
 * source vertex for each level in CSR style format. The first element of the tuple is the offset
 * array per edge type id and the second element correspond to the fanout values.
 * When heterogeneous_fan_out is provided, different fan_out values can be used for each edge type.
 * The fan-out offsets size must be proportional to the number of edge types and fan_out values.
```

Reply:
>We have no mechanism for enforcing that edge types are valid today. When the user passes in edge types, there is no parameter indicating the maximum edge type and no written assumption that edge types are densely packed and numbered from 0. The only constraint on them is that they are represented as an int32_t.

This is true for graph creation (in a way we don't require edge weights to be non-negative in graph creation), but we may better require edge types to be consecutive integers starting from 0 in heterogeneous sampling (in a way we require edge weights to be non-negative in SSSP).

https://github.com/rapidsai/cugraph/pull/4602/files#diff-2cdbfa97ec1f07c1f2c343715d235dd701988aaf562097994dcd25072816026aL596

Heterogeneous renumbering API requires `num_vertex|edge_types` to be specified and the offsets array size should be `num_vertex|edge_types + 1`. We may follow the same convention.

It is error if the graph has edge_type/vetex_type equal to or larger than `num_edge|vertex_types` and this should be checked if `do_expensive_check` is true (in a way we check whether edge weights are non-negative in SSSP if `do_expensive_check` is set to true).

Reply:
And this function handles four combinations. Uniform/Biased sampling * Homogeneous/Heterogeneous sampling. We may need to consider whether should we have one giant function handling all these combinations or create 2 or 4 separate functions. I am a bit against creating one giant function if some parameters are relevant to only a subset of combinations.

Reply:
And we can use `std::variant` to explicitly specify that a user needs to provide either `fanout` or `heterogeneous_fanout` (but we may better create a two separate functions?).

Reply:
No longer considering std::variant since both homogeneous and heterogeneous neighbor sample have the same fanout types